### ISSUE TITLE
Improve handling of concurrent requests to the same host

### ISSUE THREAD

**Issue by jogli5er** [decrypto-org/spider#18]
As of now we use an array excludedHosts, to make sure we do not get any more hosts that are already running full steam. However, this might lead to issues as noted by @dionyziz, here an excerpt from slack: 
> Another edge case is if the only remaining entries are from a particular domain.
> In that case exclusion would not allow spidering further.

In order to address this, I suggest leasing the handling of "one-level-at-a-time" scraping and allowing a higher level scrape if a request to the database to increase the dataset of the pool does not return the requested amount of entries. We then would have a maxCurrentLevel that would be allowed to scrape. Since the index is sorted by level and then by random, we would prefer lower level entries as soon as they are not excluded anymore.

**Comment by jogli5er** [comment]
If the suggested above does not work, I'll try the solution approach suggested by @dionyziz:

> We introduce randomness also to the base URL table
We also maintain a denornalized column on the base URL table which stores whether there are unexplored paths under that base
Then when we need N entries to explore, we pick N base URLs which need exploration
maintaining a composite index on the pair of columns (unexplored, randomness) for performant lookups
Then, if N domains are returned, then we take one random path for each.
If less than N are returned, we have explored all bases except N, and so we can take it slow and take one path from each of those bases returned.
In any case it will slow down but not stall completely.
And if we want to explore on a per-level basis like we're doing now, we will also need another denornalized column in the base table, minlevel
Storing the minimum level among all unexplored paths that belong to that base.
If we go that route, this can replace the unexplored column completely.
(and minlevel can take the value -1 if there are no unexplored paths under that base)
Then we index base on composite (minlevel, randomness)
And follow the same method, but for the level we are interested in

**Comment by jogli5er** [comment]
I've tested my solution shortly, and I found the following points:

- Worst case memory usage for 100 concurrent connections and a pool size of 2000 has an upper bound at 50'000 dbResults simultaneously in memory. This condition happens if we have 25 base URLs that each take up a significant amount of available paths. In this worst-case scenario, all 100 connections are busy getting data from these 25 hosts, and each time we refill the pool, we only get 2000 entries from the next, not yet overloaded host. In the next step, the previous host is excluded, and the next 2000 entries from the following host are loaded. (Which would also mean that we are unlucky with picking the random numbers)
- One dbResult has the size of about 350 bytes. Multiplied by 50'000 dbResults mentioned above in memory, we get a total of 17.5 MB.
- We did not encounter this situation when starting from our bootstrapping list; we always have plenty of choice for the pool entries.

For those reasons, I'd suggest sticking with the more straightforward solution for now.
